dao/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,237 @@
+import org.springframework.jdbc.core.simple.SimpleJdbcCall;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.math.BigDecimal;
+import java.sql.PreparedStatement;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+
+/**
+ * Default implementation of {@link com.epam.esm.dao.GiftCertificateDao} interface.
+ *
+ * @since 1.0
+ */
+@Repository
+public class GiftCertificateDaoImpl implements GiftCertificateDao {
+
+    private final static String TABLE_NAME = "gift_certificate";
+    private final static String GET_BY_ID_QUERY = "SELECT * FROM " + TABLE_NAME + " WHERE id = ?";
+    private final static String GET_ALL_QUERY = "SELECT * FROM " + TABLE_NAME;
+    private final static String SAVE_QUERY = "INSERT INTO " + TABLE_NAME + " (name, description, price, duration," +
+            " create_date, last_update_date) " + "VALUES (?,?,?,?,?,?)";
+    private final static String UPDATE_QUERY = "UPDATE " + TABLE_NAME + " SET name = ?, description = ?, " +
+            "price = ?, duration = ?, last_update_date = ? WHERE id = ?";
+    private final static String DELETE_QUERY = "DELETE FROM " + TABLE_NAME + " WHERE id = ?";
+
+    private final static String GET_ALL_BY_TAG_NAME = "SELECT gc.id, gc.name, gc.description, gc.price, gc.duration," +
+            " gc.create_date, gc.last_update_date FROM " + TABLE_NAME + " gc JOIN certificates_tags ct on gc.id = ct.certificate_id" +
+            " JOIN tag t on t.id = ct.tag_id WHERE t.name = ?";
+
+    private final static String SORTING_QUERY_FIRST_PART = "SELECT * FROM " + TABLE_NAME + " ORDER BY";
+
+    private final static String FIND_BY_NAME_PART_PROCEDURE_NAME = "searchByPartOfName";
+    private final static String FIND_BY_DESCRIPTION_PART_PROCEDURE_NAME = "searchByPartOfDescription";
+    private final static String RESULTS_SET_KEY = "certificates";
+
+    private final static String NAME_IN_PARAM_FOR_PART_OF_DESCRIPTION = "in_partOfDescription";
+    private final static String NAME_IN_PARAM_FOR_PART_OF_NAME = "in_partOfName";
+
+    private final JdbcTemplate template;
+    private final RowMapper<GiftCertificate> rowMapper;
+
+    @Autowired
+    public GiftCertificateDaoImpl(JdbcTemplate template, RowMapper<GiftCertificate> rowMapper) {
+        this.template = template;
+        this.rowMapper = rowMapper;
+    }
+
+
+    /**
+     * This method saves GiftCertificate entity.
+     * The method uses KayHolder for getting generated id for GiftCertificate entity from db.
+     *
+     * @param certificate GiftCertificate entity without id.
+     * @return GiftCertificate entity with generated id.
+     * @since 1.0
+     */
+    @Override
+    public GiftCertificate save(GiftCertificate certificate) {
+
+        String name = certificate.getName();
```

Comment:
I don't understand why are you creating all these vars and adding almost 10 more redundant lines

Reply:
comment f4bace1